### ISSUE TITLE
Maximizing Harbor network throughput

### ISSUE THREAD

**Issue by jcambronero** [goharbor/harbor#15023]
Hi, we are looking to optimize Harbor's network configuration to serve as much image pull traffic as possible.
We would like to reduce host-to-host communication in a cluster to the bare minimum required, so that all transmit network can be used for serving client pull requests.

This is what we've done so far:
- External DB and redis with plenty of resources.
- One replica of all Harbor components per host.
- Enabled service topology to serve only node local endpoints (https://kubernetes.io/docs/concepts/services-networking/service-topology/) for all Harbor ClusterIP services.
- Tried multiple options to enforce stickiness at the nginx ingress controller, using [cookie affinity](https://kubernetes.github.io/ingress-nginx/examples/affinity/cookie/) and [upstream-hash-by annotations](https://stackoverflow.com/questions/58329994/how-to-load-balance-sockets-using-ingress-nginx)

Unfortunately our tests so far show that ~30% of the host network bandwidth is still being use to serve host-to-host traffic.
We suspect that this traffic is being generated by the ingress load balancing the pull requests, since we have tested and confirmed that that the Kubernetes service topology is working as expected (i.e. requests are dropped if a node local service endpoint does not exists).

Do you have any suggestions or recommendations on what to try next?
Please let me know if there's any other helpful information I can provide.

Our setup consists of:
- Harbor v2.1.4
- Kubernetes v1.18.18

Thanks.

**Comment by renmaosheng** [comment]
Hi jcambronero , could you please elaborate the 30% percentage cost? have you done a packet capture to show this is caused by the harbor component interaction?

**Comment by jcambronero** [comment]
Thanks @renmaosheng, yes we are doing packet captures to see this behavior.
But we can more clearly see this ~30% host-to-host network communication based on the amount of _receive_ traffic network in the NIC interface when doing **pull** requests only. 

This is the scenario we are using for testing:
- Generate load testing for **pull** requests only.
- One replica of all Harbor components per host.
- Nginx ingress controller deployed as a DaemonSet.
- When we have only one host in the load balancer, we can see that the entire network bandwidth of that host is used for serving the pull requests = ~99% is for _transmit_ network traffic and only ~1% is for _receive_ network traffic
- When we add additional servers to the load balancer (i.e. to process traffic using the ingress controller), we immediately can see this host-to-host communication with ~30% of bandwidth used for  _receive_ network traffic and only ~70% for _transmit_.

**Comment by danielzhanghl** [comment]
the traffic goes from/to ingress to harbor core, then harbor registry, 
if these three pods are on different hosts,  the inter-host traffic is generated.
 just guess...

**Comment by jcambronero** [comment]
We figured it out:
- When the ingress controller uses hostPort it continues to load balance across all servers, even when using [service topology](https://kubernetes.io/docs/concepts/services-networking/service-topology/) and after trying multiple options to enforce stickiness ([cookie affinity](https://kubernetes.github.io/ingress-nginx/examples/affinity/cookie/) and [upstream-hash-by](https://stackoverflow.com/questions/58329994/how-to-load-balance-sockets-using-ingress-nginx) annotations)
- After switching the ingress controller to use NodePort with **externalTrafficPolicy: Local**, the traffic now stays locally within the host and that ~30% host-to-host overhead traffic that we saw earlier has stopped.

Just out of curiosity, does anyone have any ideas on how to achieve the same behavior that **externalTrafficPolicy: Local** produces in an ingress controller NodePort service, but using hostPort instead?

Thanks!
